src/lib/text-field/text-field-adapter.ts

```diff
@@ -1,130 +1,187 @@
-import { addClass, listenOwnProperty, getActiveElement, createElementAttributeObserver } from '@tylertech/forge-core';
+import { getShadowElement, listenOwnProperty, randomChars, toggleAttribute } from '@tylertech/forge-core';
+import { FIELD_CONSTANTS, IFieldComponent } from '../field-next';
+import { BaseFieldAdapter, IBaseFieldAdapter } from '../field-next/base/base-field-adapter';
+import { LabelComponent } from '../label';
 import { ITextFieldComponent } from './text-field';
-import { FieldAdapter, IFieldAdapter } from '../field/field-adapter';
-import { TEXT_FIELD_CONSTANTS } from './text-field-constants';
-import { FIELD_CONSTANTS } from '../field/field-constants';
-
-export interface ITextFieldAdapter extends IFieldAdapter {
-  detectTextarea(): void;
+import { TextFieldInputAttributeObserver, TextFieldValueChangeListener, TEXT_FIELD_CONSTANTS } from './text-field-constants';
+
+export interface ITextFieldAdapter extends IBaseFieldAdapter {
+  addRootListener(name: keyof HTMLElementEventMap, listener: EventListener): void;
+  disableInput(disabled: boolean): void;
+  handleDefaultSlotChange(slot: HTMLSlotElement, listener: TextFieldInputAttributeObserver): void;
+  tryAddValueChangeListener(context: unknown, listener: TextFieldValueChangeListener): void;
+  removeValueChangeListener(): void;
+  tryFloatLabel(force?: boolean): void;
+  tryConnectSlottedLabel(slot: HTMLSlotElement): void;
+  connectClearButton(listener: EventListener): void;
+  disconnectClearButton(listener: EventListener): void;
+  toggleClearButtonVisibility(visible: boolean): void;
+  clearInput(): void;
 }
 
-export class TextFieldAdapter extends FieldAdapter implements ITextFieldAdapter {
-  protected _inputElements: Array<HTMLInputElement | HTMLTextAreaElement> = [];
-  protected _inputMutationObserverInstances: MutationObserver[] = [];
+export class TextFieldAdapter extends BaseFieldAdapter implements ITextFieldAdapter {
+  protected readonly _fieldElement: IFieldComponent;
+  private readonly _clearButtonSlotElement: HTMLSlotElement;
+  private _inputElements: (HTMLInputElement | HTMLTextAreaElement)[] = [];
+  private _inputMutationObserver?: MutationObserver;
+  private _destroyValueChangerListeners: (() => void)[] = [];
+
+  public get hasValue(): boolean {
+    return this._inputElements.some(el => !!el.value);
+  }
+
+  public get hasPlaceholder(): boolean {
+    return this._inputElements.some(el => !!el.placeholder);
+  }
 
   constructor(component: ITextFieldComponent) {
     super(component);
+    this._fieldElement = getShadowElement(component, TEXT_FIELD_CONSTANTS.selectors.FIELD) as IFieldComponent;
+    this._clearButtonSlotElement = getShadowElement(component, TEXT_FIELD_CONSTANTS.selectors.CLEAR_BUTTON_SLOT) as HTMLSlotElement;
+    this._fieldElement.setAttribute('exportparts', Object.values(FIELD_CONSTANTS.parts).join(', '));
```

Comment:
Are adding the `exportparts` attr dynamically here to just reduce the noise in the static template?

Reply:
Mostly so we can define the parts just in the field constants and use them in every component that includes a field. If it weren't for that I'd prefer to add them statically, but this ends up being DRYer.